src/games/strategy/triplea/ai/proAI/util/ProOddsCalculator.java

```diff
@@ -0,0 +1,173 @@
+  }
+
+  public ProBattleResult calculateBattleResults(final PlayerID player, final Territory t,
+      final List<Unit> attackingUnits, final List<Unit> defendingUnits, final Set<Unit> bombardingUnits,
+      final boolean isAttacker) {
+
+    final ProBattleResult result = checkIfNoAttackersOrDefenders(t, attackingUnits, defendingUnits);
+    if (result != null) {
+      return result;
+    }
+    return callBattleCalculator(player, t, attackingUnits, defendingUnits, bombardingUnits);
+  }
+
+  private ProBattleResult checkIfNoAttackersOrDefenders(final Territory t, final List<Unit> attackingUnits,
+      final List<Unit> defendingUnits) {
+    final GameData data = ProData.getData();
+
+    final boolean hasNoDefenders = Match.noneMatch(defendingUnits, Matches.UnitIsNotInfrastructure);
+    final boolean isLandAndCanOnlyBeAttackedByAir = !t.isWater() && Match.allMatch(attackingUnits, Matches.UnitIsAir);
+    if (attackingUnits.size() == 0) {
+      return new ProBattleResult();
+    } else if (hasNoDefenders && isLandAndCanOnlyBeAttackedByAir) {
+      return new ProBattleResult();
+    } else if (hasNoDefenders) {
+      return new ProBattleResult(100, 0.1, true, attackingUnits, new ArrayList<Unit>(), 0);
+    } else if (Properties.getSubRetreatBeforeBattle(data) && Match.allMatch(defendingUnits, Matches.UnitIsSub)
+        && Match.noneMatch(attackingUnits, Matches.UnitIsDestroyer)) {
+      return new ProBattleResult();
+    }
+    return null;
+  }
+
+
+  public ProBattleResult callBattleCalculator(final PlayerID player, final Territory t,
+      final List<Unit> attackingUnits, final List<Unit> defendingUnits, final Set<Unit> bombardingUnits) {
+    return callBattleCalculator(player, t, attackingUnits, defendingUnits, bombardingUnits, false);
+  }
+
+  public ProBattleResult callBattleCalculator(final PlayerID player, final Territory t,
+      final List<Unit> attackingUnits, final List<Unit> defendingUnits, final Set<Unit> bombardingUnits,
+      final boolean retreatWhenOnlyAirLeft) {
+    final GameData data = ProData.getData();
+
+    if (isCanceled || attackingUnits.isEmpty() || defendingUnits.isEmpty()) {
+      return new ProBattleResult();
+    }
+
+    // Use battle calculator (hasLandUnitRemaining is always true for naval territories)
+    AggregateResults results = null;
+    final int minArmySize = Math.min(attackingUnits.size(), defendingUnits.size());
+    final int runCount = Math.max(16, 100 - minArmySize);
+    final PlayerID attacker = attackingUnits.get(0).getOwner();
+    final PlayerID defender = defendingUnits.get(0).getOwner();
+    if (retreatWhenOnlyAirLeft) {
+      calc.setRetreatWhenOnlyAirLeft(true);
+    }
+    results =
+        calc.setCalculateDataAndCalculate(attacker, defender, t, attackingUnits, defendingUnits, new ArrayList<Unit>(
+            bombardingUnits), TerritoryEffectHelper.getEffects(t), runCount);
+    if (retreatWhenOnlyAirLeft) {
```

Comment:
If that is the case, please provide diffs that make it obvious that nothing has changed. Otherwise you are not optimizing this PR for the reviewer. As is, I have to review every line to make sure that this works. You haven't provided me with anything that makes this review easier, instead of looking at it diff'ing line source A to here, that is tedious, faster for me to go through each line and make sure this still works. By that token, this is effectively brand new code, and it's dirty.